nomad/structs/structs.go

```diff
@@ -1163,6 +1215,88 @@ func ValidNodeStatus(status string) bool {
 	}
 }
 
+const (
+	// NodeSchedulingEligible and Ineligible marks the node as eligible or not,
+	// respectively, for receiving allocations. This is orthoginal to the node
+	// status being ready.
+	NodeSchedulingEligible   = "eligbile"
+	NodeSchedulingIneligible = "ineligible"
+)
+
+// DrainSpec describes a Node's desired drain behavior.
+type DrainSpec struct {
+	// Deadline is the duration after StartTime when the remaining
+	// allocations on a draining Node should be told to stop.
+	Deadline time.Duration
```

Comment:
> where is the validation logic for this? e,g make sure that deadlne is >=0

In the UpdateDrain RPC endpoint itself. A Deadline <= 0 is a force drain: https://github.com/hashicorp/nomad/pull/4010/files#diff-57d6a6426f4ff4ab5d9eecfbca4399daR446

> Its possible I missed it because of the large feature branch

I feel like it would be impossible for you to catch it in such a huge PR so no worries.